src/stateMachines/services/oid4vciMachineService.ts

```diff
@@ -0,0 +1,191 @@
+  //     })
+  //   }).then((contacts: Array<IContact>): IContact | undefined => {
+  //     // TODO throw error if 0 or more than 1
+  //     return contacts.length === 1 ? contacts[0] : undefined // TODO improve code
+  //   })
+};
+
+export const retrieveCredentials = async (
+  _ctx: OID4VCIMachineContext,
+  _event: OID4VCIMachineEventTypes,
+): Promise<Array<MappedCredentialOffer> | undefined> => {
+  const {openId4VcIssuanceProvider, verificationCode, selectedCredentials} = _ctx;
+  return openId4VcIssuanceProvider
+    ?.getCredentialsFromIssuance({
+      credentials: selectedCredentials,
+      pin: verificationCode,
+    })
+    .then(
+      (credentialOffers: Array<CredentialFromOffer>): Array<MappedCredentialOffer> =>
+        credentialOffers.map((credentialOffer: CredentialFromOffer): MappedCredentialOffer => {
+          const credentialResponse: CredentialResponse = credentialOffer.credentialResponse;
+          const verifiableCredential: W3CVerifiableCredential | undefined = credentialResponse.credential;
+          const wrappedVerifiableCredential: WrappedVerifiableCredential = CredentialMapper.toWrappedVerifiableCredential(
+            verifiableCredential as OriginalVerifiableCredential,
+          );
+          const uniformVerifiableCredential: IVerifiableCredential = wrappedVerifiableCredential.credential;
+          const rawVerifiableCredential: VerifiableCredential = credentialResponse.credential as unknown as VerifiableCredential;
+          const correlationId: string =
+            typeof uniformVerifiableCredential.issuer === 'string'
+              ? uniformVerifiableCredential.issuer
+              : (uniformVerifiableCredential.issuer as IIssuer).id;
+
+          return {
+            correlationId,
+            credentialOffer,
+            rawVerifiableCredential,
+            uniformVerifiableCredential,
+          };
+        }),
+    );
+};
+
+export const addContactIdentity = async (_ctx: OID4VCIMachineContext, _event: OID4VCIMachineEventTypes): Promise<void> => {
+  const {credentialOffers, contact} = _ctx;
+  const correlationId: string = credentialOffers[0].correlationId;
+  const identity: IBasicIdentity = {
+    alias: correlationId,
+    roles: [IdentityRoleEnum.ISSUER],
+    identifier: {
+      type: CorrelationIdentifierEnum.DID,
+      correlationId,
+    },
+  };
+  store.dispatch<any>(addIdentity({contactId: contact!.id, identity}));
+};
+
+export const verifyCredentials = async (_ctx: OID4VCIMachineContext, _event: OID4VCIMachineEventTypes): Promise<void> => {
+  const {credentialOffers, verificationCode} = _ctx;
+
+  await Promise.all(
+    credentialOffers.map(async (offer: MappedCredentialOffer): Promise<void> => {
+      const verificationResult: IVerificationResult = await verifyCredential({
+        credential: offer.credentialOffer.credentialResponse.credential as VerifiableCredential | CompactJWT, //origVC as VerifiableCredential | CompactJWT,
+        fetchRemoteContexts: true,
```

Comment:
We should probably add a todo, that we might want to allow these types of options as part of the context, now we have state machines. Allows us to pre-determine whether these policies apply and whether remote context should be fetched

Reply:
added a todo comment